internal/provider/resource_monitor_resource.go

```diff
@@ -0,0 +1,807 @@
+				Computed:    true,
+			},
+			"end_timestamp": schema.StringAttribute{
+				Description: "The date and time when the resource monitor suspends the assigned warehouses.",
+				Optional:    true,
+				Computed:    true,
+			},
+			"suspend_trigger": schema.Int64Attribute{
+				Description: "The number that represents the percentage threshold at which to suspend all warehouses.",
+				Optional:    true,
+				Validators: []validator.Int64{
+					int64validator.ConflictsWith(path.MatchRoot("suspend_triggers")),
+				},
+			},
+			"suspend_triggers": schema.SetAttribute{
+				Description: "A list of percentage thresholds at which to suspend all warehouses.",
+				Optional:    true,
+				ElementType: types.Int64Type,
+				Validators: []validator.Set{
+					setvalidator.ConflictsWith(path.MatchRoot("suspend_trigger")),
+				},
+				DeprecationMessage: "Use suspend_trigger instead",
+			},
+			"suspend_immediate_trigger": schema.Int64Attribute{
+				Description: "The number that represents the percentage threshold at which to immediately suspend all warehouses.", Optional: true,
+				Validators: []validator.Int64{
+					int64validator.ConflictsWith(path.MatchRoot("suspend_immediate_triggers")),
+				},
+			},
+			"suspend_immediate_triggers": schema.SetAttribute{
+				Description: "A list of percentage thresholds at which to suspend all warehouses.",
+				Optional:    true,
+				ElementType: types.Int64Type,
+				Validators: []validator.Set{
+					setvalidator.ConflictsWith(path.MatchRoot("suspend_immediate_trigger")),
+				},
+				DeprecationMessage: "Use suspend_immediate_trigger instead",
+			},
+			"notify_triggers": schema.SetAttribute{
+				Description: "A list of percentage thresholds at which to send an alert to subscribed users.",
+				Optional:    true,
+				ElementType: types.Int64Type,
+			},
+			"set_for_account": schema.BoolAttribute{
+				Description: "Specifies whether the resource monitor should be applied globally to your Snowflake account (defaults to false).",
+				Optional:    true,
+				Default:     booldefault.StaticBool(false),
+				// todo: create a snowflake_resource_monitor_association resource
+				// DeprecationMessage: "Use snowflake_resource_monitor_association instead",
+			},
+			"warehouses": schema.SetAttribute{
+				Description: "A list of warehouses to apply the resource monitor to.",
+				Optional:    true,
+				ElementType: types.StringType,
+				// todo: add the `resource_monitor` attribute to the `snowflake_warehouse` resource
+				// DeprecationMessage: "Set the `resource_monitor` attribute on the `snowflake_warehouse` resource instead",
+			},
+		},
+	}
+}
+
+func upgradeResourceMonitorStateV0toV1(ctx context.Context, req resource.UpgradeStateRequest, resp *resource.UpgradeStateResponse) {
+	var resourceMonitorDataV0 resourceMonitorModelV0
+	resp.Diagnostics.Append(req.State.Get(ctx, &resourceMonitorDataV0)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	name := resourceMonitorDataV0.Name
```

Comment:
We could work a bit on making this function a bit more readable. This is just a nit-picking but suggestions I have are:
- do not define all these variables upfront if they are needed in if statements much later in this function
- scope these assignments inside if condition checks
- I would like to see a builder pattern here for `resourceMonitorModelV1`; maybe not even the basic one, but the one where e.g. WithSuspendTriggers takes resourceMonitorDataV0.SuspendTriggers as an input param, ad does all the logic from if there, and finally sets the right value on resourceMonitorModelV1 struct